### ISSUE TITLE
Sanity check on CPU bottleneck with grow_gpu_hist under Windows

### ISSUE THREAD

**Issue by ghost** [dmlc/xgboost#2458]
## Environment info
Hardware:
i7 6900K @ 4.0GHz
4 x GTX 1080

Operating System:
Windows 10 Pro (x64)

Compiler:
cl.exe (Visual Studio Community 2015)

Package used (python/R/jvm/C++):
Python 3.5.2

`xgboost` version used:
0.6, compiled from source with GPU support, CUB etc. are from submodules.

If installing from source, please provide 

1. The commit hash (`git rev-parse HEAD`)
91e778c6db3f30f9b5b4ced7752a64881a0f15cb

## Steps to reproduce
```python
import xgboost as xgb
import numpy as np
from sklearn.datasets import make_classification

X, y = make_classification(1000000, n_features=50, random_state=7)
param = {'objective': 'binary:logistic', 'updater': 'grow_gpu_hist', 'n_gpus':1}
dtrain = xgb.DMatrix(X, y)
%time model = xgb.train(param, dtrain, 500)
```
Above training (using Jupyter Notebook's timing magic command, but you can use any other) takes 14.8s to run. While comparable to the 11.1s [reported](https://github.com/dmlc/xgboost/tree/master/plugin/updater_gpu) on a Pascal Titan X, the training is bottlenecked by CPU (95% CPU usage). If I set n_gpus to 4, the training finishes in 12.9s (virtually no speed up). As I don't have access to a Linux system with comparable specs right now, I just want a quick sanity check on whether the high CPU usage is a Windows specific issue, or is it WAD.

**Comment by RAMitchell** [comment]
There is a start-up overhead for the quantisation and tree algorithm set up. How many iterations did you run?

**Comment by DeepLearningJohnDoe** [comment]
500, as shown in the code. Or if you mean rerun the whole training process: no, following runs aren't any faster.

I ran the same code on a Amazon EC2 p2.xlarge (half of a K80 and 4 vCPU, and it barely got past 150% CPU utilization (equivalence of ~30% on Windows). So I am now fairly sure it wasn't supposed to happen.

**Comment by RAMitchell** [comment]
I think there is an issue with msvc. I made a note about this previously #2279.

Someone will need to profile it to find the cause. Contribution is very welcome :)

**Comment by DeepLearningJohnDoe** [comment]
I installed Ubuntu 16.04 on the same machine today and ran the same experiment.

| OS/setting | Time (s) | CPU utilization (normalized) |
| --- | --- | --- |
| Windows 10, n_gpus=1 | 14.8s | 95%
| Windows 10, n_gpus=4 | 12.9s | 95%
| Ubuntu 16.04, n_gpus=1 | 13.2s | 20%
| Ubuntu 16.04, n_gpus=4 | 46.5s (what happened here?) | 20%

**Comment by pseudotensor** [comment]
In Ubuntu, can you try adding -DCMAKE_BUILD_TYPE=Release  when calling cmake?  I think nccl is being compiled as debug.  Also, latest HEAD is smarter regarding how it handles sharding on multiple GPUs (but that doesn't explain windows vs. ubuntu behavior).  Thanks.

**Comment by pseudotensor** [comment]
@DeepLearningJohnDoe Whether more gpus will be beneficial in terms of speed depends highly upon the DMA setup (i.e. which boards have full peer to peer).  You can run tensorflow that shows the DMA map or you can run the cuda samples device inquiry code and it'll tell you the peer to peer ability across all 4 GPUs.

Speed on multiple GPUs also depends upon the data and depth.   Too small data won't be any faster as xgboost is dominated by CPU time and transfer  to/from GPU time.  Too few depth and not enough work to do on GPU.  Too large depth, and lots of communication between GPUs.  Too many features, lots of communication.  etc.

**Comment by DeepLearningJohnDoe** [comment]
I'll try -DCMAKE_BUILD_TYPE=Release come Monday.

On Windows only a specific set of cards (Everything Tesla, some Quadro, Titans up to Maxwell Titan X) has P2P access and only in TCC mode (can't drive displays); since mine are 1080s they don't have it.

**Comment by DeepLearningJohnDoe** [comment]
`-DCMAKE_BUILD_TYPE=Release` does put the wall time on Ubuntu with `n_gpus=4` at 8.2s, so I think nccl was being compiled as debug in my previous setting.

Still doesn't explain the Windows performance issue, though.

fwiw here's the output of `p2pBandwidthLatencyTest` and `nvidia-smi topo -m` under Ubuntu. Under Windows the non-diagonal terms of the connectivity matrix are zero as P2P can't be enabled.

```
P2P Connectivity Matrix
     D\D     0     1     2     3
     0	     1     1     1     1
     1	     1     1     1     1
     2	     1     1     1     1
     3	     1     1     1     1
Unidirectional P2P=Disabled Bandwidth Matrix (GB/s)
   D\D     0      1      2      3 
     0 254.48  10.09  11.21  11.19 
     1  10.07 254.81  11.27  11.29 
     2  11.28  11.28 254.64  10.08 
     3  11.26  11.28  10.08 253.65 
Unidirectional P2P=Enabled Bandwidth Matrix (GB/s)
   D\D     0      1      2      3 
     0 255.64  13.38  10.20  10.19 
     1  13.38 255.81  10.18  10.19 
     2  10.19  10.19 255.64  13.38 
     3  10.19   9.97  13.28 218.72 
Bidirectional P2P=Disabled Bandwidth Matrix (GB/s)
   D\D     0      1      2      3 
     0 256.23  10.51  19.74  19.48 
     1  10.49 256.48  19.73  19.72 
     2  19.74  19.74 256.65  10.55 
     3  19.74  19.68  10.49 213.92 
Bidirectional P2P=Enabled Bandwidth Matrix (GB/s)
   D\D     0      1      2      3 
     0 256.38  26.16  20.34  20.27 
     1  26.09 256.91  20.35  20.35 
     2  20.35  20.33 256.32  26.12 
     3  20.35  20.33  26.11 256.99 
P2P=Disabled Latency Matrix (us)
   D\D     0      1      2      3 
     0   2.36  14.69  14.44  14.11 
     1  14.34   2.30  14.68  14.23 
     2  14.35  14.63   2.34  13.95 
     3  14.27  14.90  14.66   2.31 
P2P=Enabled Latency Matrix (us)
   D\D     0      1      2      3 
     0   2.36   6.90   7.72   7.75 
     1   6.90   2.38   7.73   7.73 
     2   7.75   7.78   2.38   6.92 
     3   7.84   7.83   6.87   2.31 
```

```
	GPU0	GPU1	GPU2	GPU3	CPU Affinity
GPU0	 X 	PIX	PHB	PHB	0-15
GPU1	PIX	 X 	PHB	PHB	0-15
GPU2	PHB	PHB	 X 	PIX	0-15
GPU3	PHB	PHB	PIX	 X 	0-15
```
